tests/test_pygtracker.py

```diff
@@ -18,9 +20,124 @@ def test_version():
 
 # Start tests for generate_course_statistics
 
+
+def test_generate_course_statistics_input_not_exist():
+    """
+    Test the input course_ids are involved in MDS course list
+    """
+    tracker = generate_input_calculate_final_grade()
+    with raises(ValueError):
+        tracker.generate_course_statistics(course_ids=["500", "511"])
+
+
+def test_generate_course_statistics_input_not_str():
+    """
+    Test the input course_ids is a list of strings
+    """
+    tracker = pygtracker.GradeTracker()
+    with raises(TypeError):
+        tracker.generate_course_statistics(course_ids=[True, "511"])
+
+
+def test_generate_course_statistics_course_id_not_list_of_string():
+    """
+    Test the dtype of course_ids is a list
+    """
+    tracker = pygtracker.GradeTracker()
+    with raises(TypeError):
+        tracker.generate_course_statistics(course_ids="511")
+
+
+def test_generate_course_statistics_equal():
+    """
+    Test the function generate_course_statistics return the expected
+    results
+    """
+    tracker = generate_input_calculate_final_grade()
+    statistics = tracker.generate_course_statistics(course_ids=["511", "522"])
+    assert (
+        round(statistics.iloc[0, 1], 2) == 87.87
+    ), "The value of mean is miscalculated."
+    assert (
+        round(statistics.iloc[1, 3], 2) == 90.86
+    ), "The value of median is miscalculated."
+
+
+def test_generate_course_statistics_output_match():
+    """
+    Test the function generate_course_statistics return the expected
+    data frame
+    """
+    tracker = generate_input_calculate_final_grade()
+    row_num = len(tracker.generate_course_statistics(course_ids=["511", "522"]))
+    columns_list = tracker.generate_course_statistics(course_ids=["511", "522"]).columns
+    assert columns_list[0] == "course_id", "The first column should be course_id."
+    assert columns_list[1] == "mean", "The second column should be mean."
+    assert columns_list[2] == "1st-quantile", "The third column should be 1st-quantile."
+    assert columns_list[3] == "median", "The fourth column should be median."
+    assert columns_list[4] == "3rd-quantile", "The fifth column should be 3rd-quantile."
+    assert row_num == 2, "The row number should be 2."
+
+
 # End tests for generate_course_statistics
 
 # Start tests for rank_courses
+def test_rank_courses_input_method_type():
+    """
+    Test the dtype of method is str
+    """
+    tracker = pygtracker.GradeTracker()
```

Comment:
You should initialize this instance with data first, before trying to test that the input is wrong.

I guess if you are not providing the method, it will default to "mean", and will not trigger the exception? Can you check if this test is testing the correct scenario?

Reply:
I made a mistake here that I named the function incorrectly. I want to test the dtype of `descending` rather than `method`.
The dtype of method is not that important since we can only choose it among the four options.